File: web-api/app/schemas/printer.py
"""
Formlabs 프린터 상태 스키마
============================
Web API 응답 데이터 구조 정의
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, timezone, timedelta
from enum import Enum

# 한국 표준시 (KST = UTC+9)
KST = timezone(timedelta(hours=9))

def now_kst() -> datetime:
    """현재 한국 시간 반환"""
    return datetime.now(KST)


# ===========================================
# 열거형 정의
# ===========================================

class PrintStatus(str, Enum):
    """프린트 작업 상태"""
    QUEUED = "QUEUED"           # 대기열
    PREPRINT = "PREPRINT"       # 준비 중
    PRINTING = "PRINTING"       # 출력 중
    PAUSED = "PAUSED"           # 일시정지
    FINISHED = "FINISHED"       # 완료
    ABORTED = "ABORTED"         # 취소됨
    ERROR = "ERROR"             # 오류


class PrinterReadyState(str, Enum):
    """프린터 준비 상태"""
    READY = "READY"
    NOT_READY = "NOT_READY"
    # Formlabs API 실제 응답 값
    READY_TO_PRINT_READY = "READY_TO_PRINT_READY"
    READY_TO_PRINT_NOT_READY = "READY_TO_PRINT_NOT_READY"


class BuildPlatformState(str, Enum):
    """빌드 플랫폼 상태"""
    EMPTY = "EMPTY"
    HAS_PARTS = "HAS_PARTS"
    # Formlabs API 실제 응답 값
    BUILD_PLATFORM_CONTENTS_EMPTY = "BUILD_PLATFORM_CONTENTS_EMPTY"
    BUILD_PLATFORM_CONTENTS_HAS_PARTS = "BUILD_PLATFORM_CONTENTS_HAS_PARTS"
    BUILD_PLATFORM_CONTENTS_MISSING = "BUILD_PLATFORM_CONTENTS_MISSING"
    BUILD_PLATFORM_CONTENTS_UNCONFIRMED = "BUILD_PLATFORM_CONTENTS_UNCONFIRMED"
    BUILD_PLATFORM_CONTENTS_CONFIRMED_CLEAR = "BUILD_PLATFORM_CONTENTS_CONFIRMED_CLEAR"


# ===========================================
# 프린트 작업 스키마
# ===========================================

class CurrentPrintRun(BaseModel):
    """현재 진행 중인 프린트 작업 정보"""
    guid: Optional[str] = None
    name: Optional[str] = None
    status: Optional[PrintStatus] = None
    
    # 레이어 정보
    currently_printing_layer: int = 0
    layer_count: int = 0
    
    # 시간 정보 (밀리초)
    estimated_duration_ms: int = 0
    elapsed_duration_ms: int = 0
    estimated_time_remaining_ms: int = 0
    
    # 시작/완료 시간
    print_started_at: Optional[datetime] = None
    print_finished_at: Optional[datetime] = None
    
    @property
    def progress_percent(self) -> float:
        """진행률 (0~100)"""
        if self.layer_count == 0:
            return 0.0
        return round((self.currently_printing_layer / self.layer_count) * 100, 1)
    
    @property
    def estimated_remaining_minutes(self) -> int:
        """남은 시간 (분)"""
        return self.estimated_time_remaining_ms // 60000
    
    @property
    def elapsed_minutes(self) -> int:
        """경과 시간 (분)"""
        return self.elapsed_duration_ms // 60000


# ===========================================
# 소모품 스키마
# ===========================================

class CartridgeStatus(BaseModel):
    """레진 카트리지 상태"""
    serial: Optional[str] = None
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    initial_ml: float = 0
    remaining_ml: float = 0
    
    @property
    def remaining_percent(self) -> float:
        """잔량 퍼센트"""
        if self.initial_ml == 0:
            return 0.0
        return round((self.remaining_ml / self.initial_ml) * 100, 1)
    
    @property
    def is_low(self) -> bool:
        """잔량 부족 여부 (100ml 이하)"""
        return self.remaining_ml < 100


class TankStatus(BaseModel):
    """레진 탱크 상태"""
    serial: Optional[str] = None
    material_code: Optional[str] = None
    print_count: int = 0
    days_since_first_print: int = 0


# ===========================================
# 프린터 상태 스키마
# ===========================================

class PrinterStatus(BaseModel):
    """프린터 현재 상태"""
    status: Optional[str] = None
    last_pinged_at: Optional[datetime] = None
    current_print_run: Optional[CurrentPrintRun] = None
    ready_to_print: Optional[PrinterReadyState] = None
    build_platform_contents: Optional[BuildPlatformState] = None


class Printer(BaseModel):
    """프린터 전체 정보"""
    serial: str
    alias: Optional[str] = None  # 사용자 지정 이름
    machine_type: Optional[str] = None  # "FORM-4-0"
    
    # 상태 정보
    printer_status: Optional[PrinterStatus] = None
    cartridge_status: Optional[CartridgeStatus] = None
    tank_status: Optional[TankStatus] = None
    
    # 메타데이터
    firmware_version: Optional[str] = None
    created_at: Optional[datetime] = None
    
    @property
    def display_name(self) -> str:
        """표시용 이름 (alias 또는 serial)"""
        return self.alias or self.serial
    
    @property
    def is_printing(self) -> bool:
        """출력 중 여부"""
        if self.printer_status and self.printer_status.current_print_run:
            return self.printer_status.current_print_run.status == PrintStatus.PRINTING
        return False
    
    @property
    def is_online(self) -> bool:
        """온라인 여부 (5분 이내 통신)"""
        if self.printer_status and self.printer_status.last_pinged_at:
            now = datetime.now(timezone.utc)
            last_ping = self.printer_status.last_pinged_at
            # timezone-naive인 경우 UTC로 간주
            if last_ping.tzinfo is None:
                last_ping = last_ping.replace(tzinfo=timezone.utc)
            elapsed = now - last_ping
            return elapsed.total_seconds() < 300
        return False


# ===========================================
# 대시보드용 요약 스키마
# ===========================================

class PrinterSummary(BaseModel):
    """대시보드용 프린터 요약 정보"""
    serial: str
    name: str
    status: str  # IDLE, PRINTING, FINISHED, ERROR, OFFLINE
    
    # 출력 진행 정보 (출력 중일 때만)
    current_job_name: Optional[str] = None
    progress_percent: Optional[float] = None
    remaining_minutes: Optional[int] = None
    current_layer: Optional[int] = None
    total_layers: Optional[int] = None
    
    # 소모품 정보
    resin_remaining_ml: Optional[float] = None
    resin_remaining_percent: Optional[float] = None
    is_resin_low: bool = False
    
    # 상태 플래그
    is_online: bool = True
    is_ready: bool = True
    has_error: bool = False
    
    # 마지막 업데이트
    last_update: datetime = Field(default_factory=now_kst)


class DashboardData(BaseModel):
    """대시보드 전체 데이터"""
    printers: List[PrinterSummary]
    total_printers: int
    printers_printing: int
    printers_idle: int
    printers_error: int
    printers_offline: int
    last_update: datetime = Field(default_factory=now_kst)


# ===========================================
# 알림 스키마
# ===========================================

class NotificationType(str, Enum):
    """알림 종류"""
    PRINT_COMPLETE = "PRINT_COMPLETE"
    PRINT_ERROR = "PRINT_ERROR"
    LOW_RESIN = "LOW_RESIN"
    PRINTER_OFFLINE = "PRINTER_OFFLINE"


class Notification(BaseModel):
    """알림 정보"""
    type: NotificationType
    printer_serial: str
    printer_name: str
    title: str
    message: str
    timestamp: datetime = Field(default_factory=now_kst)
    
    # 추가 정보 (선택)
    job_name: Optional[str] = None
    error_details: Optional[str] = None


# ===========================================
# 프린트 이력 스키마
# ===========================================

class PrintHistoryPart(BaseModel):
    """프린트 이력 파트 정보"""
    display_name: str
    volume_ml: Optional[float] = None
    stl_path: Optional[str] = None  # 원본 STL 경로


class PrintHistoryItem(BaseModel):
    """프린트 이력 항목"""
    guid: str
    name: str
    printer_serial: str
    printer_name: Optional[str] = None
    status: PrintStatus

    # 시간 정보
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    duration_minutes: Optional[int] = None

    # 레이어 정보
    layer_count: int = 0

    # 재료 정보
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    estimated_ml_used: Optional[float] = None

    # 상세 정보 (확장)
    message: Optional[str] = None  # 오류/중단 메시지
    print_run_success: Optional[str] = None  # SUCCESS / null
    thumbnail_url: Optional[str] = None
    volume_ml: Optional[float] = None  # 전체 사용 레진량
    parts: List[PrintHistoryPart] = []  # 포함된 파트 목록


class PrintHistoryResponse(BaseModel):
    """프린트 이력 응답"""
    items: List[PrintHistoryItem]
    total_count: int
    page: int = 1
    page_size: int = 20

File: frontend/src/types/printer.ts
/**
 * Formlabs 프린터 타입 정의
 * 백엔드 스키마와 동기화
 */

// ===========================================
// 열거형 정의
// ===========================================

export type PrintStatus =
  | 'QUEUED'
  | 'PREPRINT'
  | 'PREHEAT'
  | 'PRECOAT'
  | 'PRINTING'
  | 'POSTCOAT'
  | 'PAUSING'
  | 'PAUSED'
  | 'ABORTING'
  | 'FINISHED'
  | 'ABORTED'
  | 'ERROR'
  | 'WAITING_FOR_RESOLUTION';

export type PrinterReadyState =
  | 'READY'
  | 'NOT_READY'
  | 'READY_TO_PRINT_READY'
  | 'READY_TO_PRINT_NOT_READY';

export type BuildPlatformState =
  | 'EMPTY'
  | 'HAS_PARTS'
  | 'BUILD_PLATFORM_CONTENTS_EMPTY'
  | 'BUILD_PLATFORM_CONTENTS_HAS_PARTS'
  | 'BUILD_PLATFORM_CONTENTS_MISSING';

export type NotificationType =
  | 'PRINT_COMPLETE'
  | 'PRINT_ERROR'
  | 'LOW_RESIN'
  | 'PRINTER_OFFLINE';

// 대시보드 표시용 상태
export type PrinterDisplayStatus =
  | 'IDLE'
  | 'PRINTING'
  | 'PREHEAT'
  | 'PAUSING'
  | 'PAUSED'
  | 'ABORTING'
  | 'FINISHED'
  | 'ERROR'
  | 'OFFLINE';

// ===========================================
// 프린트 작업
// ===========================================

export interface CurrentPrintRun {
  guid: string | null;
  name: string | null;
  status: PrintStatus | null;
  currently_printing_layer: number;
  layer_count: number;
  estimated_duration_ms: number;
  elapsed_duration_ms: number;
  estimated_time_remaining_ms: number;
  print_started_at: string | null;
  print_finished_at: string | null;
}

// ===========================================
// 소모품
// ===========================================

export interface CartridgeStatus {
  serial: string | null;
  material_code: string | null;
  material_name: string | null;
  initial_ml: number;
  remaining_ml: number;
}

export interface TankStatus {
  serial: string | null;
  material_code: string | null;
  print_count: number;
  days_since_first_print: number;
}

// ===========================================
// 프린터 상태
// ===========================================

export interface PrinterStatus {
  status: string | null;
  last_pinged_at: string | null;
  current_print_run: CurrentPrintRun | null;
  ready_to_print: PrinterReadyState | null;
  build_platform_contents: BuildPlatformState | null;
}

export interface Printer {
  serial: string;
  alias: string | null;
  machine_type: string | null;
  printer_status: PrinterStatus | null;
  cartridge_status: CartridgeStatus | null;
  tank_status: TankStatus | null;
  firmware_version: string | null;
  created_at: string | null;
}

// ===========================================
// 대시보드용 요약
// ===========================================

export interface PrinterSummary {
  serial: string;
  name: string;
  status: PrinterDisplayStatus;
  current_job_name: string | null;
  progress_percent: number | null;
  remaining_minutes: number | null;
  elapsed_minutes: number | null;
  estimated_total_minutes: number | null;
  current_layer: number | null;
  total_layers: number | null;
  print_started_at: string | null;
  print_phase: string | null;
  temperature: number | null;
  resin_remaining_ml: number | null;
  resin_remaining_percent: number | null;
  is_resin_low: boolean;
  cartridge_material_code: string | null;
  cartridge_material_name: string | null;
  machine_type: string | null;
  firmware_version: string | null;
  tank_serial: string | null;
  tank_material_code: string | null;
  tank_print_count: number | null;
  is_online: boolean;
  is_ready: boolean;
  ready_to_print: string | null;
  build_platform_contents: string | null;
  has_error: boolean;
  last_update: string;
}

export interface DashboardData {
  printers: PrinterSummary[];
  total_printers: number;
  printers_printing: number;
  printers_idle: number;
  printers_error: number;
  printers_offline: number;
  last_update: string;
}

// ===========================================
// 알림
// ===========================================

export interface Notification {
  type: NotificationType;
  printer_serial: string;
  printer_name: string;
  title: string;
  message: string;
  timestamp: string;
  job_name: string | null;
  error_details: string | null;
}

// ===========================================
// 프린트 이력
// ===========================================

export interface PrintHistoryPart {
  display_name: string;
  volume_ml: number | null;
  stl_path: string | null;
}

export interface PrintHistoryItem {
  guid: string;
  name: string;
  printer_serial: string;
  printer_name: string | null;
  status: PrintStatus;
  started_at: string | null;
  finished_at: string | null;
  duration_minutes: number | null;
  layer_count: number;
  material_code: string | null;
  material_name: string | null;
  estimated_ml_used: number | null;
  // 상세 정보 (확장)
  message: string | null;
  print_run_success: string | null;
  thumbnail_url: string | null;
  volume_ml: number | null;
  parts: PrintHistoryPart[];
}

export interface PrintHistoryResponse {
  items: PrintHistoryItem[];
  total_count: number;
  page: number;
  page_size: number;
}

// ===========================================
// WebSocket 메시지
// ===========================================

export interface WebSocketMessage {
  type: 'dashboard_update' | 'notification' | 'printer_update';
  data: DashboardData | Notification | PrinterSummary;
  timestamp: string;
}

// ===========================================
// 유틸리티 함수
// ===========================================

export function getStatusColor(status: PrinterDisplayStatus): string {
  switch (status) {
    case 'PRINTING':
      return 'blue';
    case 'PREHEAT':
      return 'orange';
    case 'PAUSING':
    case 'PAUSED':
      return 'yellow';
    case 'ABORTING':
      return 'red';
    case 'FINISHED':
      return 'green';
    case 'IDLE':
      return 'gray';
    case 'ERROR':
      return 'red';
    case 'OFFLINE':
      return 'gray';
    default:
      return 'gray';
  }
}

export function getStatusLabel(status: PrinterDisplayStatus): string {
  switch (status) {
    case 'PRINTING':
      return '출력 중';
    case 'PREHEAT':
      return '예열 중';
    case 'PAUSING':
      return '일시정지 중';
    case 'PAUSED':
      return '일시정지됨';
    case 'ABORTING':
      return '중단 중';
    case 'FINISHED':
      return '완료';
    case 'IDLE':
      return '대기 중';
    case 'ERROR':
      return '오류';
    case 'OFFLINE':
      return '오프라인';
    default:
      return '알 수 없음';
  }
}

export function formatDuration(minutes: number): string {
  if (minutes < 60) {
    return `${minutes}분`;
  }
  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;
  return mins > 0 ? `${hours}시간 ${mins}분` : `${hours}시간`;
}

export function formatResinAmount(ml: number | null): string {
  if (ml === null) return '-';
  return `${ml.toFixed(0)}ml`;
}
